src/core/server/saved_objects/migrationsv2/next.ts

```diff
@@ -63,19 +61,28 @@ export const nextActionMap = (client: ElasticsearchClient, transformRawDocs: Tra
       Actions.createIndex(client, state.targetIndex, state.targetIndexMappings),
     CREATE_REINDEX_TEMP: (state: CreateReindexTempState) =>
       Actions.createIndex(client, state.tempIndex, state.tempIndexMappings),
-    REINDEX_SOURCE_TO_TEMP: (state: ReindexSourceToTempState) =>
-      Actions.reindex(
+    REINDEX_SOURCE_TO_TEMP_OPEN_PIT: (state: ReindexSourceToTempOpenPit) =>
+      Actions.openPit(client, state.sourceIndex.value),
+    REINDEX_SOURCE_TO_TEMP_READ: (state: ReindexSourceToTempRead) =>
+      Actions.readWithPit(
         client,
-        state.sourceIndex.value,
+        state.sourceIndexPitId,
+        state.unusedTypesQuery,
+        state.batchSize,
+        state.lastHitSortValue
+      ),
+    REINDEX_SOURCE_TO_TEMP_CLOSE_PIT: (state: ReindexSourceToTempClosePit) =>
+      Actions.closePit(client, state.sourceIndexPitId),
+    REINDEX_SOURCE_TO_TEMP_INDEX: (state: ReindexSourceToTempIndex) =>
+      Actions.transformDocs(
+        client,
+        transformRawDocs,
+        state.outdatedDocuments,
         state.tempIndex,
-        Option.none,
-        false,
-        state.unusedTypesQuery
+        false
```

Comment:
I'm not sure where's the best place to put such a comment, probably in the README and somewhere in the code.

Since we're not waiting for a refresh anymore, it's quite important that _something_ waits for a refresh before we get to `OUTDATED_DOCUMENTS_SEARCH`. `UPDATE_TARGET_MAPPINGS`/`pickupUpdatedMappings` forces a refresh so that will ensure that everything works, but it could be easy to introduce a bug later if we decide to remove this step. Maybe we actually need an end-to-end integration test for this scenario.

Reply:
> Maybe we actually need an end-to-end integration test for this scenario.

@rudolf Isn't it covered by existing migration integration tests?

Reply:
Actually thinking through it again I realise the risk isn't that high. If we don't refresh then `OUTDATED_DOCUMENTS_SEARCH` will return one batch of results (even though all documents were already up to date), but after waiting for that batch to be written we will have waited for the refresh to finish. So worst case we introduce one unnecessary OUTDATED_DOCUMENTS_SEARCH/WRITE step.

We won't pick up this case in the integration test, since it will succeed, just be slightly slower, but since this is very low impact I don't think it's worth testing for it anymore.